File: examples/k8s-sandbox/kafka-throughput/consumer/consumer.py
import logging
import os
import time

from kafka import KafkaConsumer  # type: ignore
from kafka.errors import NoBrokersAvailable  # type: ignore

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)

# Configuration from environment variables
KAFKA_BROKER = os.environ.get("KAFKA_BROKER", "kafka:9092")
KAFKA_TOPIC = os.environ.get("KAFKA_TOPIC", "throughput-topic")
CONSUMER_GROUP_ID = os.environ.get(
    "CONSUMER_GROUP_ID", "kafka-throughput-consumer-group"
)
CONSUMER_CLIENT_ID = os.environ.get("CONSUMER_CLIENT_ID", "kafka-throughput-consumer")
MAX_RETRIES = int(os.environ.get("MAX_RETRIES", "5"))
INITIAL_RETRY_DELAY_S = int(os.environ.get("INITIAL_RETRY_DELAY_S", "5"))


def main() -> None:
    logging.info("--- Kafka Throughput Demo Consumer (Simplified) --- ")
    logging.info(f"Target Kafka Broker: {KAFKA_BROKER}")
    logging.info(f"Target Kafka Topic: {KAFKA_TOPIC}")
    logging.info(f"Consumer Group ID: {CONSUMER_GROUP_ID}")
    logging.info(f"Client ID: {CONSUMER_CLIENT_ID}")

    consumer = None
    retries = 0
    while retries < MAX_RETRIES:
        try:
            consumer = KafkaConsumer(
                KAFKA_TOPIC,
                bootstrap_servers=KAFKA_BROKER,
                group_id=CONSUMER_GROUP_ID,
                client_id=CONSUMER_CLIENT_ID,
                auto_offset_reset="latest",
            )
            logging.info(
                "KafkaConsumer initialized successfully. Waiting for messages..."
            )
            break
        except NoBrokersAvailable as e:
            retries += 1
            logging.warning(
                f"Attempt {retries}/{MAX_RETRIES}: Could not connect to Kafka broker "
                f"at {KAFKA_BROKER}. Error: {e}"
            )
            if retries >= MAX_RETRIES:
                logging.error("Maximum connection attempts reached. Exiting.")
                exit(1)
            delay = INITIAL_RETRY_DELAY_S * (2 ** (retries - 1))
            logging.info(f"Retrying in {delay} seconds...")
            time.sleep(delay)
        except Exception as e:
            logging.error(
                f"An unexpected error occurred during KafkaConsumer initialization: {e}"
            )
            exit(1)

    if not consumer:
        logging.error("Failed to initialize Kafka consumer after multiple retries.")
        exit(1)

    message_counter = 0
    log_interval = 1000  # Log every 1000 messages for example

    try:
        for message in consumer:
            message_counter += 1
            if message_counter % log_interval == 0:
                logging.info(
                    f"Consumed {message_counter} messages. "
                    f"Last offset: {message.offset}"
                )

    except KeyboardInterrupt:
        logging.info("Consumer interrupted. Shutting down...")
    except Exception as e:
        logging.error(f"An unexpected error occurred during consumption: {e}")
    finally:
        if consumer:
            logging.info("Closing Kafka consumer...")
            consumer.close()
            logging.info("Consumer closed.")
        logging.info(f"Total messages processed by this instance: {message_counter}")


if __name__ == "__main__":
    main()

File: examples/k8s-sandbox/kafka-throughput/consumer/consumer.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"github.com/IBM/sarama"
)

// Configuration from environment variables with defaults
var (
	kafkaBroker         = getEnv("KAFKA_BROKER", "kafka:9092")
	kafkaTopic          = getEnv("KAFKA_TOPIC", "throughput-topic")
	consumerGroupID     = getEnv("CONSUMER_GROUP_ID", "kafka-throughput-consumer-group")
	consumerClientID    = getEnv("CONSUMER_CLIENT_ID", "kafka-throughput-consumer")
	maxRetries          = getEnvAsInt("MAX_RETRIES", 5)
	initialRetryDelayS  = getEnvAsInt("INITIAL_RETRY_DELAY_S", 5)
	logInterval         = 5000 // Messages between log entries
)

func getEnv(key, defaultValue string) string {
	if value, exists := os.LookupEnv(key); exists {
		return value
	}
	return defaultValue
}

func getEnvAsInt(key string, defaultValue int) int {
	if value, exists := os.LookupEnv(key); exists {
		if intVal, err := strconv.Atoi(value); err == nil {
			return intVal
		}
	}
	return defaultValue
}

func setupLogger() *log.Logger {
	return log.New(os.Stdout, "", log.LstdFlags)
}

func createConsumerGroup() (sarama.ConsumerGroup, error) {
	config := sarama.NewConfig()
	config.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategyRoundRobin
	config.Consumer.Offsets.Initial = sarama.OffsetNewest
	config.ClientID = consumerClientID

	// Retry logic for connecting to Kafka
	var consumerGroup sarama.ConsumerGroup
	var err error
	for retry := 0; retry < maxRetries; retry++ {
		consumerGroup, err = sarama.NewConsumerGroup([]string{kafkaBroker}, consumerGroupID, config)
		if err == nil {
			return consumerGroup, nil
		}

		log.Printf("Attempt %d/%d: Could not connect to Kafka broker at %s. Error: %v",
			retry+1, maxRetries, kafkaBroker, err)

		if retry < maxRetries-1 {
			delay := time.Duration(initialRetryDelayS*(1<<retry)) * time.Second
			log.Printf("Retrying in %v...", delay)
			time.Sleep(delay)
		}
	}
	return nil, fmt.Errorf("failed to connect after %d attempts: %v", maxRetries, err)
}

type Consumer struct {
	logger *log.Logger
	ready  chan bool
	msgCount int64
}

func (c *Consumer) Setup(_ sarama.ConsumerGroupSession) error {
	close(c.ready)
	return nil
}

func (c *Consumer) Cleanup(_ sarama.ConsumerGroupSession) error {
	return nil
}

func (c *Consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
	for message := range claim.Messages() {
		c.msgCount++
		if c.msgCount%int64(logInterval) == 0 {
			c.logger.Printf("Consumed %d messages. Last offset: %d", c.msgCount, message.Offset)
		}
		session.MarkMessage(message, "")
	}
	return nil
}

func main() {
	logger := setupLogger()
	logger.Println("--- Kafka Throughput Demo Consumer (Go) ---")
	logger.Printf("Target Kafka Broker: %s", kafkaBroker)
	logger.Printf("Target Kafka Topic: %s", kafkaTopic)
	logger.Printf("Consumer Group ID: %s", consumerGroupID)
	logger.Printf("Client ID: %s", consumerClientID)

	consumerGroup, err := createConsumerGroup()
	if err != nil {
		logger.Fatalf("Error creating consumer group: %v", err)
	}
	defer consumerGroup.Close()

	ctx, cancel := context.WithCancel(context.Background())
	consumer := &Consumer{
		logger: logger,
		ready:  make(chan bool),
	}

	// Handle graceful shutdown
	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-signals
		logger.Println("Consumer interrupted. Shutting down...")
		cancel()
	}()

	// Start consuming
	for {
		topics := []string{kafkaTopic}
		err := consumerGroup.Consume(ctx, topics, consumer)
		if err != nil {
			logger.Printf("Error from consumer: %v", err)
		}
		if ctx.Err() != nil {
			break
		}
		consumer.ready = make(chan bool)
	}

	logger.Printf("Total messages processed by this instance: %d", consumer.msgCount)
}
